### ISSUE TITLE
type.models and observable array dont work together

### ISSUE THREAD

**Issue by ditiem** [mobxjs/mobx-state-tree#1176]
**_Bug report_**

* [ x] I've checked documentation and searched for existing issues
* [ x] I've made sure my project is based on the latest MST version

**Sandbox link or minimal reproduction code**
https://codesandbox.io/s/48mn350jv7

Or the code, if you prefer to paste it in https://mattiamanzati.github.io/mobx-state-tree-playground/ :
```
import React from "react"
import { types } from "mobx-state-tree"
import { observable } from 'mobx'
import { observer } from "mobx-react"
import { inspect, render } from "mobx-state-tree-playground"

const AppModel = types.model({
    count: types.optional(types.number, 10)
}).actions(self => ({
    increment(){
        self.count++
    },
	decrement() {
		self.count--
	}
}))

const ArrM = types.model( { arr: types.array( AppModel ) }) ;

const ddbb = AppModel.create()
const cache = ArrM.create( { arr: [ ddbb ] } )
const a = observable( [ cache[ 0 ] ] )
inspect(store)

const App = observer(
    props => <div>
        My awesome counter: 
        <button onClick={() => props.store[0].decrement()}>-</button> 
        {props.store[0].count} 
        <button onClick={() => props.store[0].increment()}>+</button> 
    </div>
)

render(<App store={a} />)
```

**Describe the expected behavior**
I have a types.model instance (cache in the code) that consists in an array of other types.model (cache = [ ddbb ] ) instances. Getting a sublist (in this case [ cache[ 0 ] ])  and storing it in a simple observable array (I called it a) do end in throwing exception when attempting to read the value: Cannot read property 'count' of undefined. This is thrown in the mobx getter function.

**Describe the observed behavior**
To read the value without any exception.

**Comment by xaviergonz** [comment]
Don't you mean
`const a = observable([cache.arr[0]])`

**Comment by ditiem** [comment]
Indeed!

With`const a = observable([cache.arr[0]])` things works as expected, which means that the bug I found is not so easy to reproduce. I have been investigating, trying to write the same code in codesandbox in js (my project is in TS) and the only differences I found are:
- It may happen only in Typescript
- In my code, the variable `a` is a class field named `langs`. So this is what has amazed me. This way of defining `langs` makes the code to fail:
```@observable public langs: any[] = []```
But in this way, it works as expected:
```public langs: any[] = observable( [] )```

I tried to make codesandbox to use typescript but I did not succeed. I will try later to see if I am able to make it work, otherwise I am afraid I will close the issue :(

**Comment by xaviergonz** [comment]
I'd check the js generated in both cases and compare

**Comment by mweststrate** [comment]
Please throw away your node_modules, and check if there are warnings during npm / yarn install. You might be trying to bundle incompatible versions. Also check your browser console, to see if there are any warnings / errors. Finally, at least always share the stack trace if you have one, so that we have at least something to go by.

**Comment by ditiem** [comment]
Give me a couple of days and I will post it, I am rewriting the system and now it is all "broken".

**Comment by ditiem** [comment]
I have recorder a video where the bug happens. It is just before and after the assign. It happens with all the fields, not just `langs`.

This is how lang is declared in the class:

```
class MyClass
{
  ...
  @observable public langs              : EnumLangType[]
  ...
```

in the assignment, `langs` is a list of types.model( ).create( ).

Here is backtrace when I access `langs` from the console (that is the same that I see on the screen when the program executes for a bit longer.

```
> this.langs[ 0 ].desc
mobx.module.js:884 Uncaught TypeError: Cannot read property 'read' of undefined
    at Proxy.get (mobx.module.js:4063)
    at ComputedValue$$1../node_modules/mobx/lib/mobx.module.js.ComputedValue$$1.computeValue (mobx.module.js:942)
    at ComputedValue$$1../node_modules/mobx/lib/mobx.module.js.ComputedValue$$1.get (mobx.module.js:872)
    at ObservableObjectAdministration$$1../node_modules/mobx/lib/mobx.module.js.ObservableObjectAdministration$$1.read (mobx.module.js:3822)
    at Object.get (mobx.module.js:4086)
    at Object.get (mobx.module.js:2774)
    at eval (eval at <anonymous> (20_model_system.ts:247), <anonymous>:1:17)
    at TourOperator.<anonymous> (20_model_system.ts:247)
    at step (17_model_quotation.ts:877)
    at Object.next (17_model_quotation.ts:877)
    at 17_model_quotation.ts:877
    at new Promise (<anonymous>)
    at ./src/03_model/20_model_system.ts.__awaiter (17_model_quotation.ts:877)
    at 20_model_system.ts:225
get @ mobx.module.js:4063
./node_modules/mobx/lib/mobx.module.js.ComputedValue$$1.computeValue @ mobx.module.js:942
./node_modules/mobx/lib/mobx.module.js.ComputedValue$$1.get @ mobx.module.js:872
./node_modules/mobx/lib/mobx.module.js.ObservableObjectAdministration$$1.read @ mobx.module.js:3822
get @ mobx.module.js:4086
get @ mobx.module.js:2774
...
```

Versions:
```
    "mobx": "5.9.0",
    "mobx-react": "5.3.6",
    "mobx-state-tree": "^3.10.2",
```

Any light is super welcome, I am really stuck

[MST-bug.m4v.zip](https://github.com/mobxjs/mobx-state-tree/files/2873284/MST-bug.m4v.zip)

**Comment by xaviergonz** [comment]
Just so you know the array given by mst is already observable so it should be OK to remove the observable decorator in this case.
Either way, it looks like mobx is trying to make the observable mst an observable object and screwing things up...

**Comment by xaviergonz** [comment]
Nah, that works, just checked.

This works too :-S
```
test("#1176", () => {
    class X {
        @observable
        public arr: any[] = []
    }
    const x = new X()

    const M = types.model({
        arr: types.array(
            types.model({
                a: types.string
            })
        )
    })
    const m = M.create({
        arr: [{ a: "hi" }]
    })
    x.arr = m.arr
    expect(x.arr[0].a).toBe("hi")
})
```

**Comment by ditiem** [comment]
The array given is not "observable". I had no choice other than removing the "observable", but now my code does not react.

I suspect mobx is screwing it up because is trying to convert the array and its members to an observable, so the "getters" and "setters" are overwritten. 

The code is like this (it is what I tried to reproduce in the first message, but I could not):

```
const lang_model = types.model( { ... } )
const ddbb_model = types.model( { langs: types.map( lang_model ) } ) ;

... some insertions take place in ddbb_model.langs ...

const langs = [ ddbb_model.langs[ 'es' ], ddbb_model.langs[ 'en' ] ] ;

... later in the class ...
this.langs = langs ... boom!
```

**Comment by ditiem** [comment]
I think I am starting to understand what is going on. "Objects" of MST of type `{$treenode: ObjectNode$$1, …}` are transformed into mobx object of type `Proxy {Symbol(mobx administration): ObservableObjectAdministration$$1}`. This happens when a MST object is inserted/added into a MobX observer, or when some Data Structure (read Map, Array, etc) that is not an observable contains MST nodes that are transformed into observable.

Does this trigger any idea?

**Comment by xaviergonz** [comment]
Maybe you got two copies of mobx for some reason and one cannot figure out the other is already observable and it's trying to make it so?

**Comment by xaviergonz** [comment]
I'd delete node modules and yarn.lock and run yarn again just to make sure

**Comment by ditiem** [comment]
I just did it. All clean, installed and rebuild from scratch. Same :(

BUT your previous message gave me a hint. I use ant design (antd) package, which I have seen uses mobx in one of its projects.

Could that influence? (they use 5.1.0 and I do use 5.9.0)

and if it does, do you have any idea about how can I detect the "duality" of mobx-* libraries?

**Comment by xaviergonz** [comment]
It absolutely could, since they use a symbol to detect when something is already observable and both symbols would be different. I think it should show a warning in the console when multiple versions are detected.

More info https://github.com/mobxjs/mobx/issues/1082

**Comment by ditiem** [comment]
Thank you Xavier!

There is a second copy of mobx in the repository!!! after I removed by hand everything seems to work as expected.

But catching it has not been easy. The reason is that I have some local packages that I have linked using `yarn link` to the main project or to another of these local packages. Yarn seems to create a filesystem link in the node_modules directory to the linked package, and there is where the second copy of mobx was (the linked package contains another node_modules directory). Not easy to find as a find do not follow symlinks by default (need to add -L). No "warning message" appear in the console.

Thank you for all your help!
